### PR TITLE
Add ability to pass custom uploader methods

### LINKED ISSUE #19: Add request handlers
Can you add `handlers` to endpoints.
If you look at the backend side, yes, it's easier to just specify the url and line up the backend with the format you want. But if you look at it from the frontend side, it is easier to manage the process of sending a file to the backend. 
For cases when graphql is used as a transport instead of rest. 
For cases when it is more convenient to use your api layer than to adjust to your api.
Something like that
```js
  {
    handlers: {
      byFile (file, done) {
        /// magic code
        done(url)
      },
      byUrl (url, done) {
        /// magic code
        done(urlFromBackend)
      },
    },
  }

// or promise
{
      handlers: {
        async byFile (file) {
          // magic code
          return url;
        },
        async byUrl (url) {
          // magic code
          return urlFromBackend;
        },
      },
}
```

### LINKED ISSUE #24: Allow to execute a function when File is uploaded, along with passing the Values...
I think it would be great if you just allow the **calling a function** passing the image data as parameter and then will use it to send to backend with custom configs (if the user wants to edit some config). That will make it more efficient for developers to use it...

### PR DESCRIPTION
**Pull request by neSpecc**
Resolves #19
Resolves #24 

Co-Authored-By: dimensi <eddimensi@gmail.com>

### CODE REVIEW

**Comment by neSpecc** [comment]
Wait for `codex.ajax` library update with the new `selectFiles` method

**Comment by neSpecc** [comment]
Resolves #24

File: README.md [hunk 1]
```diff
@@ -77,13 +84,16 @@ Image Tool supports these configuration parameters:
...
+| endpoints | `{byFile: string, byUrl: string}` | Endpoints for file uploading. <br> Contains 2 fields: <br> __byFile__ - for file uploading <br> __byUrl__ - for uploading by URL |
 | field | `string` | (default: `image`) Name of uploaded image field in POST request |
 | types | `string` | (default: `image/*`) Mime-types of files that can be [accepted with file selection](https://github.com/codex-team/ajax#accept-string).|
 | additionalRequestData | `object` | Object with any data you want to send with uploading requests |
 | additionalRequestHeaders | `object` | Object with any custom headers which will be added to request. [See example](https://github.com/codex-team/ajax/blob/e5bc2a2391a18574c88b7ecd6508c29974c3e27f/README.md#headers-object) |
 | captionPlaceholder | `string` | (default: `Caption`) Placeholder for Caption input |
 | buttonContent | `string` | Allows to override HTML content of «Select file» button |
+| uploader | `{{uploadByFile: function, uploadByUrl: function}}` | Optional custom uploading methods. See details below. |
+
+Note that if you won't implement your custom uploader methods, so the `endpoints` param is required.
```
Comment (gohabereg) [inline README.md]:
```suggestion
Note that if you don't implement your custom uploader methods, the `endpoints` param is required. 
```

File: src/uploader.js [hunk 2]
```diff
@@ -24,27 +24,50 @@ export default class Uploader {
...
-          onPreview(e.target.result);
-        };
-      },
-      fieldName: this.config.field
-    })
-      .then((response) => {
-        this.onUpload(response);
-      })
-      .catch((error) => {
+    let preparePreview = function (file) {
```
Comment (gohabereg) [inline src/uploader.js]:
```suggestion
    const preparePreview = function (file) {
```

File: src/uploader.js [hunk 3]
```diff
@@ -87,30 +126,52 @@ export default class Uploader {
...
 
-    formData.append(this.config.field, file);
+      if (!isPromise(upload)) {
+        console.warn('Custom uploader method uploadByFile should return a Promise');
+      }
+    } else {
+      /**
+       * Default uploading
+       */
+      let formData = new FormData();
```
Comment (gohabereg) [inline src/uploader.js]:
```suggestion
      const formData = new FormData();
```

**Pull request merged**